### ISSUE TITLE
Does not work on k8s version 1.22

### ISSUE THREAD

**Issue by helloqiu** [tikv/tikv-operator#15]
Hi there. K8s released version 1.22 in which, according to [their official blogs](https://kubernetes.io/blog/2021/08/04/kubernetes-1-22-release-announcement/#removal-of-several-deprecated-beta-apis), they removed the deperacated CRD API `apiextensions.k8s.io/v1beta1`. Do you have plans to release a new CRD file with `apiextensions.k8s.io/v1`?

**Comment by maoueh** [comment]
Here resources I used to make it work on Kubernetes 1.22:

```
# For Kubernetes >= 1.16.
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: tikvclusters.tikv.org
spec:
  group: tikv.org
  scope: Namespaced
  names:
    plural: tikvclusters
    singular: tikvcluster
    kind: TikvCluster
  versions:
  - name: v1alpha1
    served: true
    storage: true
    subresources:
      status: {}
    schema:
      openAPIV3Schema:
        type: object
        properties:
          spec:
            type: object
            properties:
              version:
                type: string
              pd:
                type: object
                properties:
                  baseImage:
                    type: string
                  replicas:
                    type: number
                  storageClassName:
                    type: string
                    nullable: true
                  requests:
                    type: object
                    properties:
                      storage:
                        type: string
                  config:
                    type: object
              tikv:
                type: object
                properties:
                  baseImage:
                    type: string
                  replicas:
                    type: number
                  storageClassName:
                    type: string
                    nullable: true
                  requests:
                    type: object
                    properties:
                      storage:
                        type: string
                  config:
                    type: object
          status:
            type: object
            properties:
              pd:
                type: object
                properties:
                  synced:
                    type: boolean
                  phase:
                    type: string
                  statefulSet:
                    type: object
                  members:
                    type: object
                    additionalProperties:
                      type: object
                      properties:
                        name:
                          type: string
                        id:
                          type: string
                        clientURL:
                          type: string
                        health:
                          type: boolean
                        lastTransitionTime:
                          type: string
                          format: datetime
                  leader:
                    type: object
                    properties:
                      name:
                        type: string
                      id:
                        type: string
                      clientURL:
                        type: string
                      health:
                        type: boolean
                      lastTransitionTime:
                        type: string
                        format: datetime
                  failureMembers:
                    type: object
                    additionalProperties:
                      type: object
                      properties:
                        podName:
                          type: string
                        memberID:
                          type: string
                        pvcUID:
                          type: string
                        memberDeleted:
                          type: boolean
                        createdAt:
                          type: string
                          format: datetime
                  unjoinedMembers:
                    type: object
                    additionalProperties:
                      type: object
                      properties:
                        podName:
                          type: string
                        pvcUID:
                          type: string
                        createdAt:
                          type: string
                          format: datetime
                  image:
                    type: string
              tikv:
                type: object
                properties:
                  synced:
                    type: boolean
                  phase:
                    type: string
                  statefulSet:
                    type: object
                  stores:
                    type: object
                    additionalProperties:
                      type: object
                      properties:
                        id:
                          type: string
                        podName:
                          type: string
                        ip:
                          type: string
                        leaderCount:
                          type: integer
                          format: int32
                        state:
                          type: string
                        lastHeartbeatTime:
                          type: string
                          format: datetime
                        lastTransitionTime:
                          type: string
                          format: datetime
                  tombstoneStores:
                    type: object
                    additionalProperties:
                      type: object
                      properties:
                        id:
                          type: string
                        podName:
                          type: string
                        ip:
                          type: string
                        leaderCount:
                          type: integer
                          format: int32
                        state:
                          type: string
                        lastHeartbeatTime:
                          type: string
                          format: datetime
                        lastTransitionTime:
                          type: string
                          format: datetime
                  failureStores:
                    type: object
                    additionalProperties:
                      type: object
                      properties:
                        podName:
                          type: string
                        storeID:
                          type: string
                        createdAt:
                          type: string
                          format: datetime
                  image:
                    type: string
              conditions:
                type: array
                items:
                  type: object
                  properties:
                    type:
                      type: string
                    status:
                      type: string
                    lastUpdateTime:
                      type: string
                      format: datetime
                    lastTransitionTime:
                      type: string
                      format: datetime
                    reason:
                      type: string
                    message:
                      type: string
    additionalPrinterColumns:
    - jsonPath: .status.conditions[?(@.type=="Ready")].status
      name: Ready
      type: string
    - jsonPath: .status.pd.image
      description: The image for PD cluster
      name: PD
      type: string
    - jsonPath: .spec.pd.replicas
      description: The desired replicas number of PD cluster
      name: PD Desire
      type: integer
    - jsonPath: .status.pd.statefulSet.readyReplicas
      description: The current replicas number of PD cluster
      name: Current
      type: integer
    - jsonPath: .status.tikv.image
      description: The image for TiKV cluster
      name: TiKV
      type: string
    - jsonPath: .spec.tikv.replicas
      description: The desired replicas number of TiKV cluster
      name: TiKV Desire
      type: integer
    - jsonPath: .status.tikv.statefulSet.readyReplicas
      description: The current replicas number of TiKV cluster
      name: Current
      type: integer
    - jsonPath: .metadata.creationTimestamp
      name: Age
      type: date
    - jsonPath: .status.conditions[?(@.type=="Ready")].message
      name: Status
      priority: 1
      type: string
```

And the actual cluster definitions:

```
# IT IS NOT SUITABLE FOR PRODUCTION USE.
# This YAML describes a basic TiDB cluster with minimum resource requirements,
# which should be able to run in any Kubernetes cluster with storage support.
apiVersion: tikv.org/v1alpha1
kind: TikvCluster
metadata:
  name: basic
spec:
  version: v4.0.0
  pd:
    baseImage: pingcap/pd
    replicas: 1
    # if storageClassName is not set, the default Storage Class of the Kubernetes cluster will be used
    # storageClassName: local-storage
    requests:
      storage: "1Gi"
    config: {}
  tikv:
    baseImage: pingcap/tikv
    replicas: 1
    # if storageClassName is not set, the default Storage Class of the Kubernetes cluster will be used
    # storageClassName: local-storage
    requests:
      storage: "1Gi"
    config: {}
```

### CRD 

I written down some important part of the schema, however, it's definitely not exhaustive. It seems tools exists to extract from Go the actual definition and then convert them to `YAML` so it can be simply copy/paste.

 I used https://github.com/kubernetes/kube-openapi inside this repository and were able to generate Go definition of the OpenAPI schema with:
 
 ```
 openapi-gen --go-header-file ./hack/boilerplate/boilerplate.go.txt -i github.com/tikv/tikv-operator/pkg/apis/tikv/v1alpha1 --output-package=generated
 ```
 
 I stopped there not knowing how to turn that generated output into a usable YAML definition for `versions.schema.openAPIV3Schema`.
 
 It seems that other tools can be used to generate them, I read about https://github.com/kubernetes-sigs/controller-tools and more specifically there `controller-gen` tool. I did not investigate further.

**Comment by maoueh** [comment]
### Problems

It seems the custom status `subresources` and the actual `additionalPrinterColumns` do not work properly. I tried to debug a big to understand why it's not working, did not find the actual problem.

**Comment by jan-xyz** [comment]
Hey @maoueh I fixed a very similar problem for an etcd-operator just recently. The PR to it should include some steps on how to go about it: https://github.com/cbws/etcd-operator/pull/419/files

TL;DR: I generated the YAML definitions, included these in the binary and applied the resources during CRD creation by reading them from disc.
